include/k52/optimization/steepest_descent_method.h

```diff
@@ -0,0 +1,120 @@
+public:
+	/// @brief Creates instance of HookeJeevesMethod.
+	/// @param alfa - number that defines optimal multiplication step for Golden Section Search method
+	/// @param increment_of_the_argument - number that defines value of gradient step per coordinate
+	/// @param max_iteration_number - maximum number of iterations before computations will be stopped
+	/// @param precision - number that defines accuracy of our computations
+	SteepestDescentMethod(
+		double alpha = 1.6,
+		double increment_of_the_argument = 0.00001,
+		size_t max_iteration_number = 1000,
+		double precision = 0.0001);
+
+	/// @brief  See ICloneable.
+	/// Creates deep clone of an object with resource allocation.
+	/// @return deep clone of an object
+	virtual SteepestDescentMethod* Clone() const;
+
+	virtual std::string get_name() const;
+
+#ifdef BUILD_WITH_MPI
+	virtual void Send(boost::mpi::communicator* communicator, int target) const;
+
+	virtual void Receive(boost::mpi::communicator* communicator, int source);
+#endif
+
+protected:
+	virtual std::vector<double> FindOptimalParameters(const std::vector<double>&);
+
+	/// @brief Checks if steps vector length satisfies an exit criteria.
+	/// @param steps_array - vector of steps values for diffirent coordinates 
+	/// @return true if steps vector norm less than specified precision, false other way
+	bool IsExitCriteriaFulfilled(
+		std::vector<double>& steps_array) const;
+
+	/// @brief Calculates gradient in the point.
+	/// @param parameters - vector of coordinates
+	/// @return new vector with direction to the function increase
+	std::vector<double> CalculateGradient(
+		const std::vector<double>& parameters) const;
+
+	/// @brief Calculates partial derivative for the point.
+	/// @param parameters - vector of coordinates
+	/// @param index - index of coordinate
+	/// @return value of partial derivative for the point
+	double CalculateDerivative(
+		const std::vector<double>& parameters,
+		size_t& index) const;
+
+	/// @brief Multiply each coordinate of vector by -1
+	/// @param parameters - vector of coordinates
+	/// @return invert vector
+	std::vector<double> InvertVector(
+		std::vector<double>& parameters) const;
+
+	/// @brief Calculate optimal coefficient for Steepest descent method
+	/// @param init_point - vector of current point
+	/// @param step - previous step in Steepest descent method
+	/// @param direction - vector of steps for each coordiante
+	/// @return cooficient for next step in Steepest descent method
+	double GoldenSectionSearch(
+		std::vector<double>& init_point, 
+		double& step, 
+		std::vector<double>& direction) const;
+
+	/// @brief Supporting method for search of Steepest descent optimal coefficient
+	/// @param init_parameters - vector of current point
+	/// @param step_for_borders - step for localization segment with minimum of function
+	/// @param direction - vector of steps for each coordiante
+	/// @return cooficient for next step in Steepest descent method
+	double Find_lambda(
```

Comment:
Lambda - it's length of step in direction, I rename this in FindStepLength, it's be normal?